src/DotNetWorker.Grpc/Handlers/InvocationHandler.cs

```diff
@@ -0,0 +1,159 @@
+                return response;
+            }
+
+            try
+            {
+                var invocation = new GrpcFunctionInvocation(request);
+
+                IInvocationFeatures invocationFeatures = _invocationFeaturesFactory.Create();
+                invocationFeatures.Set<FunctionInvocation>(invocation);
+                invocationFeatures.Set<IExecutionRetryFeature>(invocation);
+
+                context = _application.CreateContext(invocationFeatures, cancellationTokenSource.Token);
+                invocationFeatures.Set<IFunctionBindingsFeature>(new GrpcFunctionBindingsFeature(context, request, _outputBindingsInfoProvider));
+
+                if (_inputConversionFeatureProvider.TryCreate(typeof(DefaultInputConversionFeature), out var conversion))
+                {
+                    invocationFeatures.Set<IInputConversionFeature>(conversion!);
+                }
+
+                await _application.InvokeFunctionAsync(context);
+
+                var functionBindings = context.GetBindings();
+
+                foreach (var binding in functionBindings.OutputBindingData)
+                {
+                    var parameterBinding = new ParameterBinding
+                    {
+                        Name = binding.Key
+                    };
+
+                    if (binding.Value is not null)
+                    {
+                        parameterBinding.Data = await binding.Value.ToRpcAsync(_serializer);
+                    }
+
+                    response.OutputData.Add(parameterBinding);
+                }
+
+                if (functionBindings.InvocationResult is not null)
+                {
+                    TypedData? returnVal = await functionBindings.InvocationResult.ToRpcAsync(_serializer);
+                    response.ReturnValue = returnVal;
+                }
+
+                response.Result.Status = StatusResult.Types.Status.Success;
+            }
+            catch (Exception ex)
+            {
+                response.Result.Exception = ex.ToRpcException();
+                response.Result.Status = StatusResult.Types.Status.Failure;
+
+                if (ex.InnerException is TaskCanceledException or OperationCanceledException)
+                {
+                    response.Result.Status = StatusResult.Types.Status.Cancelled;
+                }
+            }
+            finally
+            {
+                _inflightInvocations.TryRemove(request.InvocationId, out var cts);
+
+                if (context is IAsyncDisposable asyncContext)
+                {
+                    await asyncContext.DisposeAsync();
+                }
+
+                (context as IDisposable)?.Dispose();
+            }
+
+            return response;
+        }
+
+        public bool TryCancel(string invocationId)
+        {
+            _inflightInvocations.TryGetValue(invocationId, out var cancellationTokenSource);
+
+            if (cancellationTokenSource is CancellationTokenSource)
```

Comment:
not blocking nit: it is nice to see defensive coding, but this can be simplified since you know you never add a `null` CTS to the dictionary:

``` CSharp
if (_inflightInvocations.TryGetValue(invocationId, out var cancellationTokenSource))
{
    try
    {
        cancellationTokenSource.Cancel();
        return true;
    }
    catch (ObjectDisposedException)
    {
    }
    catch (Exception ex)
    {
        _logger.LogWarning(ex, "Unable to cancel invocation {invocationId}.", invocationId);
    }
}

return false;
```